### PR TITLE
add golang bindings for v2 implementation of preflight

### PR DESCRIPTION
**Pull request by OchiengEd**
This PR is a work in progress and aims to be used in v2 of preflight.

It checks for the existence of both podman and docker sockets and uses the available socket to interact with containers and images.

The current implementation only supports a handful of methods. More methods need to be implemented for instance to allow exporting container filesystems and running commands in a container

More design and work needs to be done to integrate with the existing code and to provide a mechanism to run operator-sdk / scorecard tests.

See example.go file in the bindings package for an idea how to interact with the tools.

Comments and contributions are welcome.

**TODO:**
- write unit tests

**EXAMPLE:**

```
import (
 	"fmt"
)

func TestListImages() {
	client, err := Client()
	if err != nil {
		fmt.Println("error:", err.Error())
	}
	images, err := client.ListImages()
	if err != nil {
		fmt.Println(err.Error())
	}

	fmt.Printf("%+v/n", images)
}
```

### CODE REVIEW

**Comment by OchiengEd** [comment]
let me know if a rebase is needed

File: certification/internal/bindings/types.go
Comment (bcrochet) [inline certification/internal/bindings/types.go:42]: Most of these methods should have a return type of a struct specific to that operation (PullImageReport, ListImageReport, etc.). It would "future proof" the return type a bit, so that as more structured data is returned, existing code wouldn't necessarily need to change.

File: certification/internal/bindings/podman.go
Comment (bcrochet) [inline certification/internal/bindings/podman.go:21]: Can we at least name this with the nameOrID?

**Review by bcrochet — CHANGES_REQUESTED** [review]
Couple of comments to start.

File: certification/internal/bindings/podman.go
Comment (OchiengEd) [reply certification/internal/bindings/podman.go:21]: What are you referring to as "this"? I have no reference point.

File: certification/internal/bindings/types.go
Comment (OchiengEd) [reply certification/internal/bindings/types.go:42]: roger that

File: certification/internal/bindings/podman.go
Comment (bcrochet) [reply certification/internal/bindings/podman.go:21]: image.tar.gz is an extremely generic name. That filename should be created based on nameOrID.

{nameOrID}.tar.gz

**Comment by OchiengEd** [comment]
@bcrochet recommendations done.

File: certification/internal/bindings/example.go
Comment (komish) [inline certification/internal/bindings/example.go:7]: Can you go ahead and remove this file, and instead just add the example to a comment on this PR?

File: certification/internal/bindings/helpers.go
Comment (komish) [inline certification/internal/bindings/helpers.go:61]: Add a comment describing the behavior

File: certification/internal/bindings/helpers.go
Comment (komish) [inline certification/internal/bindings/helpers.go:73]: Add a comment describing the behavior

**Review by komish — CHANGES_REQUESTED** [review]
A couple of changes in-line.

This PR gets us closer to solving the problem of supporting both docker and podman environments. It does, however, require that we rewrite some of our checks in order to make use of it.

I also think we'll have to introduce a few client methods that we're needing for bundle checks that are not represented here.

Finally, this handles containerCLI calls, but doesn't handle skopeo/operator-sdk based calls. We'll need to add those for the checks that currently utilize them in a future PR.

I approve merging based on the fact that this:
1) doesn't impact the existing code execution
2) moves us closer to supporting more developer workstations

From that perspective, we'll need to create additional issues that explores utilizing these bindings as a replacement for our current container CLI invocations, and writing the tests to support these new bindings.

@OchiengEd if you can create these issues, that'd be great. Otherwise, I'll create them after this gets merged.

File: certification/internal/bindings/helpers.go
Comment (komish) [reply certification/internal/bindings/helpers.go:61]: The reason I ask for this is because the input and output types appear similar in their description (the parameter name, and the return type), so the comment would help clarify to us the exact reason for this function existing without having to understand the code's behavior. This also applies to the next location.

**Comment by komish** [comment]
@OchiengEd I also see a test failing for what appears to be a missing header file for a dependency? Not sure if we need to tidy up the gomod or adjust the test environment.

File: certification/internal/bindings/helpers.go
Comment (OchiengEd) [reply certification/internal/bindings/helpers.go]: this function has been removed.

File: certification/internal/bindings/example.go
Comment (OchiengEd) [reply certification/internal/bindings/example.go]: this has been done. Been added to the body of the PR.

File: certification/internal/bindings/helpers.go
Comment (OchiengEd) [reply certification/internal/bindings/helpers.go:66]: done.

**Comment by OchiengEd** [comment]
> @OchiengEd I also see a test failing for what appears to be a missing header file for a dependency? Not sure if we need to tidy up the gomod or adjust the test environment.

a rebase was needed to fix this.

**Review by bcrochet — COMMENTED** [review]
A quick comment. If you don't agree, I'm fine with it.

**Review by bcrochet — APPROVED** [review]

File: certification/internal/bindings/docker.go
Comment (OchiengEd) [reply certification/internal/bindings/docker.go:31]: I think this approach is better as opposed to every author of a test using this function choosing a different location to save the archive. With that also comes the ease to override.

However, if we determine there is a good reason to pass the parameter later, I would be open to that. This approach ensures all our tarballs are in the same location.

**Comment by komish** [comment]
I'm looking into the failing tests. I think it's a dependency problem in the environment used for GH Actions. I'll re-review and approve once I've sorted that out.

**Comment by komish** [comment]
I believe the fix in PR https://github.com/redhat-openshift-ecosystem/openshift-preflight/pull/135 should resolve the failing tests here. After that merges, we should see passing tests here.

**Pull request merged**
